### PR TITLE
docs: Add documentation on how to submit PRs

### PR DESCRIPTION
**Pull request by p12tic**
http://trac.buildbot.net/wiki/SubmittingPatches is out of date, is hard to update and hosts part Buildbot documentation separately from the rest. This PR moves it to Buildbot documentation proper, updates it and adds precise information on how to run various tests locally.

### CODE REVIEW

File: master/docs/developer/pull-request.rst [hunk 1]
```diff
@@ -0,0 +1,212 @@
+Submitting Pull Requests
+========================
+
+As Buildbot it is used by software developers, it tends to receive a significant number of patches.
```
Comment (seankelly) [inline master/docs/developer/pull-request.rst]: Drop the first "it".

File: master/docs/developer/pull-request.rst [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+This assumes some familiarity with git, but not too much. Note that GitHub has some great `​Git guides <http://github.com/guides>`_ to get you started.
+
+Guidelines
+----------
+
+* Pull requests should be based on the latest development code, not on the most recent release.
+  That is, you should check out the `master` branch and develop on top of it.
+
+* Pull requests should include code changes, relevant documentation changes, and relevant unit tests.
+  Any patch longer than a few lines which does not have documentation or tests will probably be rejected with a request to add documentation or tests.
```
Comment (seankelly) [inline master/docs/developer/pull-request.rst]: I think I would rephrase it to not use "rejected". It can feel like the PR will be closed immediately when instead we ask for these things and the PR waits until docs and tests are updated.

Perhaps:
>   Any patch longer than a few lines which does not have documentation or tests is unlikely to be merged.

File: master/docs/developer/pull-request.rst [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+* Python code in Buildbot uses four-space indentations, with no tabs.
+  Lines should be wrapped before the 100th column.
+
+* Pull requests must reliably pass all tests.
+  Buildbot does not tolerate "flaky" tests.
+  If you have trouble with tests that fail without any of your changes applied, get in touch with the developers for help.
+
+* Pull requests that add features or change existing behavior should include a brief description in the release notes.
+  See the master/buildbot/newsfragments directory and read the README.md file therein.
```
Comment (seankelly) [inline master/docs/developer/pull-request.rst]: Can we link to this?

File: master/docs/developer/pull-request.rst [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+* Git commit messages form the "ChangeLog" for Buildbot, and as such should be as descriptive as possible.
+
+* Backward and forward compatibility is important to Buildbot. Try to minimize the effect of your patch on existing users.
```
Comment (seankelly) [inline master/docs/developer/pull-request.rst]: Second sentence snuck into this line.

File: master/docs/developer/pull-request.rst [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+Additional suggestions
+~~~~~~~~~~~~~~~~~~~~~~
+
+The Buildbot developers are quite busy, and it can take a while to review a patch.
+While the following are not required, they will make things easier for you and the developers:
+
+* Make a distinct pull request, on a distinct branch in your repository, for each unrelated change.
+  Some pull request may get merged immediately, while others will require revision, and this can get very confusing in a single branch.
+
+* Smaller, incremental commits are better than one large commits, as they can be considered on their own merits.
```
Comment (seankelly) [inline master/docs/developer/pull-request.rst]: "one large commit"*

File: master/docs/developer/pull-request.rst [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+  Be careful to credit others for their work, if you include it in your code.
+
+How to create a pull request
+----------------------------
+
+.. note::
+
+   See ​`this github guide <https://help.github.com/en/articles/fork-a-repo>`_ which offers a more generic description of this process.
+
+* Sign up for a free account at ​http://github.com, if you don't already have one.
```
Comment (seankelly) [inline master/docs/developer/pull-request.rst]: https* here and next line.

File: master/docs/developer/pull-request.rst [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+* Go to ​http://github.com/buildbot/buildbot and click “fork”.
+  This will create your own public copy of the latest buildbot source.
```
Comment (seankelly) [inline master/docs/developer/pull-request.rst]: "Buildbot"*

File: master/docs/developer/pull-request.rst [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+
+* Clone your forked repository on your local machine, so you can do your hacking.
```
Comment (seankelly) [inline master/docs/developer/pull-request.rst]: I would use "modifications" or "changes" instead of "hacking".

File: master/docs/developer/pull-request.rst [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+Not all test suites have been documented so far, only these that fail most often.
+Before each of the commands detailed below, a virtualenv must be setup as described in :ref:`PythonDevQuickStart`:
+
+.. code-block:: bash
+
+    make virtualenv
+    . .venv/bin/activate
+
+If you see weird test results after changing branches of the repository, remove the `.venv` directory and repeat above again.
+Note that `pip install -r <file>.txt` only need to be run once at the beginning of your testing session.
```
Comment (seankelly) [inline master/docs/developer/pull-request.rst]: "needs"*

File: master/docs/developer/pull-request.rst [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+
+Master unit tests
+~~~~~~~~~~~~~~~~~
+
+Tests in this category run the Python unit tests for the master.
+These tests are represented by **bb/trial/** test names in the BuildBot CI.
```
Comment (seankelly) [inline master/docs/developer/pull-request.rst]: "Buildbot"* and for subsequent paragraphs.

File: master/docs/developer/pull-request.rst [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+To run locally, execute the following:
+
+.. code-block:: bash
+
+    pip install -r requirements-ci.txt
+    trial -j16 buildbot  # change -j parameter to fit the number of cores you have
```
Comment (seankelly) [inline master/docs/developer/pull-request.rst]: We should use a more conservative `-j` value because many will run it verbatim.

Comment (p12tic) [reply master/docs/developer/pull-request.rst]: Replaced with -j8. Even the weakest machines now seem to have 2 cores with 4 hyperthreads, developer machines are likely to have 4 cores with 8 hyperthreads.

**Comment by p12tic** [comment]
@seankelly Thanks for proofreading. I think I fixed all the issues.
